File: 백준/Gold/7579. 앱/앱.py
import sys

def si():
    return sys.stdin.readline().rstrip()


# dp[i][j] -> i번째 인덱스의 앱까지 고려했을때 바용 j 만큼 사용하여 얻을 수 있는 최대메모리
def preprocess():
    # Tabulation
    for i in range(1, n + 1):
        for j in range(10001):
            if cost[i] > j:
                dp[i][j] = dp[i - 1][j]
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - cost[i]] + memory[i])


def solution():
    for j in range(10001):
        if dp[n][j] >= m:
            print(j)
            return


n, m = map(int, si().split())
memory = [0] + list(map(int, si().split()))
cost = [0] + list(map(int, si().split()))

# dp[n][m] -> i번째 인덱스까지의 앱으로 메모리 j만큼 확보 할때 필요한 최소 가치(비용)
dp = [[0 for _ in range(10001)] for _ in range(n + 1)]

preprocess()
solution()

File: 백준/Gold/7579. 앱/앱.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        int[] memory = new int[N + 1];
        int[] cost = new int[N + 1];
        int[][] dp = new int[N + 1][10001];

        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= N; i++) {
            memory[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= N; i++) {
            cost[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = 1; i <= N; i++) {
            for (int j = 0; j <= 10000; j++) {
                if (i == 1) {
                    if (j - cost[i] >= 0) dp[i][j] = memory[i];
                } else {
                    if (j - cost[i] >= 0) {
                        dp[i][j] = Math.max(dp[i - 1][j - cost[i]] + memory[i], dp[i - 1][j]);
                    } else {
                        dp[i][j] = dp[i - 1][j];
                    }
                }
            }
        }


        for (int j = 0; j <= 10000; j++) {
            if (dp[N][j] >= M) {
                System.out.println(j);
                return;
            }
        }
    }
}

